packages/@o3r/schematics/src/rule-factories/ng-add/index.ts

```diff
@@ -7,15 +7,25 @@ import { NgAddPackageOptions, NodePackageNgAddTask } from '../../tasks/index';
  * @param packages List of packages to be installed via `ng add`
  * @param options install options
  */
-export function ngAddPackages(packages: string[], options?: NgAddPackageOptions) {
-  return (tree: Tree, context: SchematicContext): Tree => {
+export function ngAddPackages(packages: string[], options?: NgAddPackageOptions): Rule {
+  const getInstalledVersion = async (packageName: string) => {
+    try {
+      return (await import(`${packageName}/package.json`)).version;
+    } catch (e) {
+      return;
+    }
+  };
+  return async (_tree: Tree, context: SchematicContext): Promise<Rule> => {
     if (packages.length > 0) {
       context.logger.info(`'${options?.parentPackageInfo || ''}' - 'ng add' has been launched for the following packages:`);
-      packages.forEach((packageName) => {
-        context.logger.info(`Running ng add for: ${packageName}${options?.version ? ' with version: ' + options.version : ''}`);
-        return context.addTask(new NodePackageNgAddTask(packageName, options));
-      });
+      for (const packageName of packages) {
+        const installedVersion = await getInstalledVersion(packageName);
+        if (!installedVersion || options?.version !== installedVersion) {
+          context.logger.info(`Running ng add for: ${packageName}${options?.version ? ' with version: ' + options.version : ''}`);
+          context.addTask(new NodePackageNgAddTask(packageName, options));
+        }
+      }
     }
-    return tree;
+    return noop;
```

Comment:
It is fine but why not returning the tree anymore?

Reply:
for some reason a `Rule` cannot return `Promise<Tree>` (even if that seems to work at runtime), it accepts `Promise<Rule | void>`